include/aspect/material_model/rheology/strain_dependent.h

```diff
@@ -0,0 +1,221 @@
+      {
+        public:
+          /**
+           * Declare the parameters this function takes through input files.
+           */
+          static
+          void
+          declare_parameters (ParameterHandler &prm);
+
+          /**
+           * Read the parameters from the parameter file.
+           */
+          void
+          parse_parameters (ParameterHandler &prm);
+
+          /**
+           * A function that computes how the rheologic parameters change
+           * if strain weakening is applied. Given a compositional field with
+           * the index j and a vector of all compositional fields, it returns
+           * the weakened cohesion, friction angle and a reduction factor for
+           * the prefactor of the viscous flow law(s) used in the computation
+           * for that composition.
+           */
+          std::array<double, 3>
+          compute_weakened_yield_parameters(const unsigned int j,
+                                            const std::vector<double> cohesions,
+                                            const std::vector<double> angles_internal_friction,
+                                            const std::vector<double> &composition) const;
+
+          /**
+           * A function that computes the strain weakened values
+           * of cohesion and internal friction angle for a given
+           * compositional field.
+           */
+          std::pair<double, double>
+          calculate_plastic_weakening (const double strain_ii,
+                                       const std::vector<double> cohesions,
+                                       const std::vector<double> angles_internal_friction,
+                                       const unsigned int j) const;
+
+          /**
+           * A function that computes by how much the diffusion and dislocation
+           * prefactors for a given compositional field are weakened under the
+           * influence of a given strain.
+           */
+          double
+          calculate_viscous_weakening (const double strain_ii,
+                                       const unsigned int j) const;
+
+          /**
+           * A function that fills the reaction terms for the finite strain tensor in
+           * MaterialModelOutputs object that is handed over. It assumes the first
+           * component of the finite strain tensor is named 's11' and all other
+           * components follow this compositional field.
+           */
+          void compute_finite_strain_reaction_terms (const MaterialModel::MaterialModelInputs<dim> &in,
+                                                     MaterialModel::MaterialModelOutputs<dim> &out) const;
+          /**
+           * A function that fills the reaction terms for the finite strain invariant(s) in
+           * MaterialModelOutputs object that is handed over.
+           */
+          void fill_reaction_outputs (const MaterialModel::MaterialModelInputs<dim> &in,
```

Comment:
Perspectively, I would like to move this somewhere else (maybe make another sub-structure like `reactions`), but this would require to pass the necessary information around (like what type of strain should be tracked). So for now lets keep it together with the rest of the finite strain stuff like you did here.